ccmlib/node.py

```diff
@@ -426,19 +426,16 @@ def watch_log_for(self, exprs, from_mark=None, timeout=600, process=None, verbos
                 line = f.readline()
                 if line:
                     reads = reads + line
-                    for e in tofind:
+                    search = tofind
+                    for e in search:
                         m = e.search(line)
                         if m:
                             matchings.append((line, m))
-
-                            # This is bogus! - One should not modify the list from inside the loop which iterates on
-                            # that list.
-                            # However since we are going to break from the loop a few lines below that should be ok.
                             tofind.remove(e)
 
-                            if len(tofind) == 0:
-                                return matchings[0] if isinstance(exprs, string_types) else matchings
-                            break
+                    if len(tofind) == 0:
+                        return matchings[0] if isinstance(exprs, string_types) else matchings
+                    break
```

Comment:
This break previously applied to the inner search loop, now it will break from outer loop on the first line, no?

Reply:
Yes. The break marks that we've found all expected matches. Instead of iterating and modifying the same set, we now copy the search and use "tofind" (original) only as carrier of matches remaining.

Reply:
No, wait, now I got your question. This looks wrong. Hang on...